README.md
# Geditor
This gem distills from gblog, my old light weight blog. it includes light weight texteditor, textile like syntax and code syntax highlight.

# Installation

Add this line to your application's Gemfile:

    gem 'geditor'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install geditor

## Usage

in view for generating text:
<%= geditor_buttons %>
<%= f.text_area :content, :id => "geditor" %>

in view for display text:
<%= g_format content %>

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

CHANGELOG.md
## v0.0.1

* initial release

## v0.0.2

* fix button image bug
* use better helper name: geditor_buttons

## v0.0.3

* fix bug when puts js in page bottom
* remove stupid svn info

lib/javascripts/geditor.js
/*****************************************/
// Name: Javascript Textarea HTML Editor
// Version: 1.3
// Author: Balakrishnan
// Modified by: Gary Zhang
// Last Modified Date: 13/Sep/2012
// License: Free
// URL: http://www.corpocrat.com
/******************************************/

var textarea;
var content;
function gEditor(obj) {
	document.write("<img class=\"button\" src=\"/assets/geditor/bold.png\" name=\"btnBold\" title=\"Bold\" onClick=\"doAddTags('*','*','" + obj + "')\">");
	document.write("<img class=\"button\" src=\"/assets/geditor/underline.png\" name=\"btnUnderline\" title=\"Underline\" onClick=\"doAddTags('+','+','" + obj + "')\">");
	document.write("<img class=\"button\" src=\"/assets/geditor/h1.png\" name=\"btnHead\" title=\"Head1\" onClick=\"doAddTags('[h]','[/h]','" + obj + "')\">");
	document.write("<img class=\"button\" src=\"/assets/geditor/ordered.png\" name=\"btnList\" title=\"Unordered List\" onClick=\"doList('<ul>','</ul>','" + obj + "')\">");
  	document.write("<img class=\"button\" src=\"/assets/geditor/code.png\" name=\"btnCode\" title=\"Code\" onClick=\"doAddTags('[code:ruby]','[/code]','" + obj + "')\">");
	document.write("<img class=\"button\" src=\"/assets/geditor/link.png\" name=\"btnLink\" title=\"Insert Link\" onClick=\"doAddTags('[link: link_description]','[/link]','" + obj + "')\">");
	document.write("<img class=\"button\" src=\"/assets/geditor/image.png\" name=\"btnPicture\" title=\"Insert Picture\" onClick=\"doAddTags('[img]','[/img]','" + obj + "')\">");
	document.write("<br>"); }
		
function doAddTags(tag1,tag2,obj)
{
	textarea = document.getElementById(obj);
	if (document.selection) // Code for IE
	  {
		  textarea.focus();
			var sel = document.selection.createRange();
			sel.text = tag1 + sel.text + tag2;
		}
   else // Code for Mozilla Firefox
     {
			var len = textarea.value.length;
		   var start = textarea.selectionStart;
			var end = textarea.selectionEnd;
			var scrollTop = textarea.scrollTop;
			var scrollLeft = textarea.scrollLeft;
        var sel = textarea.value.substring(start, end);
			var rep = tag1 + sel + tag2;
        textarea.value =  textarea.value.substring(0,start) + rep + textarea.value.substring(end,len);		
			textarea.scrollTop = scrollTop;
			textarea.scrollLeft = scrollLeft;
		}
}

function doList(tag1,tag2,obj)
{
	textarea = document.getElementById(obj);	
	if (document.selection) // Code for IE
	{
	  textarea.focus();
		var sel = document.selection.createRange();
		var list = sel.text.split('\n');
		for(i=0;i<list.length;i++) 
		{ list[i] = '* ' + list[i]; }
		sel.text = list.join("\n");
	} 
	else // Code for Firefox
	{
		var len = textarea.value.length;
	   var start = textarea.selectionStart;
		var end = textarea.selectionEnd;
		var i;
		var scrollTop = textarea.scrollTop;
		var scrollLeft = textarea.scrollLeft;
     var sel = textarea.value.substring(start, end);
		var list = sel.split('\n');
		for(i=0;i<list.length;i++) 
		{ list[i] = '* ' + list[i]; }
		var rep = list.join("\n");
		textarea.value =  textarea.value.substring(0,start) + rep + textarea.value.substring(end,len);
		textarea.scrollTop = scrollTop;
		textarea.scrollLeft = scrollLeft;
 	 }
}

lib/models/gformat_string.rb
class GformatString < String
  def to_html_format
    content = self
    classes_in_order = self.class.subclasses.map(&:to_s).sort.map(&:constantize)
    classes_in_order.each do |formatable_string|
      content = formatable_string.new(content).format
    end
    return content
  end
end

#importent to put it in first (not good)
class ListString < GformatString
  def format
    content = self + "\n" # add \n
    content.gsub!(/(^\* [^\n]+\n)+/) {|ul| "<ul>#{ul.gsub!(/\* [^\n]+\n/){|li| "<li>#{li[2..-2]}</li>"} }<\/ul>"}
    return content[0..-2] # remove \n
  end
end

class NString < GformatString
  def format
    return self.gsub("\n", "<br/>\n\r").gsub("\t", "&nbsp;&nbsp;")
  end
end

class FontSizeString < GformatString
  def format
    return self.gsub(/\[h\](.+)\[\/h\]/,'<span style="font-size:27px">\1</span>')
  end
end

class ImageString < GformatString
  def format
    return self.gsub(/\[img(:*.*)\](.+?)\[\/img\]/,'<img style="float\1" src="\2">')
  end
end

class UrlString < GformatString
  def format
    return self.gsub(/\[link:(.+?)\](.+?)\[\/link\]/, '<a target="_blank" style="color:#2894FF" href=\2>\1</a>')
  end
end

class RuleString < GformatString
  RULES = {"*" => "</b>", "#" => "</h>"}
  def format
    @content = self
    unescape_rule = generate_unescape_rule
    escape_rule = generate_escape_rule(unescape_rule)
    escape(escape_rule)
    convert
    escape(unescape_rule)
    return @content
  end

  private
  def convert
    RULES.each do |key, value|
      while @content =~ /#{"\\"+key}(.+?)#{"\\"+key}/m
        @content = $` + value.delete("/") + $1 + value + $'
      end
    end
  end

  def generate_unescape_rule
     un_escape_rule = []
      RULES.each_key do |key|
        un_escape_rule << [Kernel.rand.to_s, key]
      end
    un_escape_rule
  end

  def generate_escape_rule(un_escape_rule)
    return un_escape_rule.map{|rule| ["\\" + rule[1], rule[0]]}
  end

  def escape(escape_rules)
    escape_rules.each { |escape_rule| @content.gsub!(escape_rule[0], escape_rule[1]) }
  end
end

lib/helpers/geditor_helper.rb
module GeditorHelper
  def g_format(text)
    if text
      result = ""
      while text =~ /(\[code\:([a-z].+?)\](.+?)\[\/code\])/m
        result << GformatString.new($`).to_html_format + $&
        text = $'
      end
      result << GformatString.new(text).to_html_format
      return raw CoderayString.new(result).format
    end
  end
  
  def geditor_buttons
    raw %q(
    <script>
    /*****************************************/
    // Name: Javascript Textarea HTML Editor
    // Version: 1.3
    // Author: Balakrishnan
    // Modified by: Gary Zhang
    // Last Modified Date: 13/Sep/2012
    // License: Free
    // URL: http://www.corpocrat.com
    /******************************************/

    var textarea;
    var content;
    function gEditor(obj) {
    	document.write("<img class=\"button\" src=\"/assets/geditor/bold.png\" name=\"btnBold\" title=\"Bold\" onClick=\"doAddTags('*','*','" + obj + "')\">");
    	document.write("<img class=\"button\" src=\"/assets/geditor/underline.png\" name=\"btnUnderline\" title=\"Underline\" onClick=\"doAddTags('+','+','" + obj + "')\">");
    	document.write("<img class=\"button\" src=\"/assets/geditor/h1.png\" name=\"btnHead\" title=\"Head1\" onClick=\"doAddTags('[h]','[/h]','" + obj + "')\">");
    	document.write("<img class=\"button\" src=\"/assets/geditor/ordered.png\" name=\"btnList\" title=\"Unordered List\" onClick=\"doList('<ul>','</ul>','" + obj + "')\">");
      	document.write("<img class=\"button\" src=\"/assets/geditor/code.png\" name=\"btnCode\" title=\"Code\" onClick=\"doAddTags('[code:ruby]','[/code]','" + obj + "')\">");
    	document.write("<img class=\"button\" src=\"/assets/geditor/link.png\" name=\"btnLink\" title=\"Insert Link\" onClick=\"doAddTags('[link: link_description]','[/link]','" + obj + "')\">");
    	document.write("<img class=\"button\" src=\"/assets/geditor/image.png\" name=\"btnPicture\" title=\"Insert Picture\" onClick=\"doAddTags('[img]','[/img]','" + obj + "')\">");
    	document.write("<br>"); }
		
    function doAddTags(tag1,tag2,obj)
    {
    	textarea = document.getElementById(obj);
    	if (document.selection) // Code for IE
    	  {
    		  textarea.focus();
    			var sel = document.selection.createRange();
    			sel.text = tag1 + sel.text + tag2;
    		}
       else // Code for Mozilla Firefox
         {
    			var len = textarea.value.length;
    		   var start = textarea.selectionStart;
    			var end = textarea.selectionEnd;
    			var scrollTop = textarea.scrollTop;
    			var scrollLeft = textarea.scrollLeft;
            var sel = textarea.value.substring(start, end);
    			var rep = tag1 + sel + tag2;
            textarea.value =  textarea.value.substring(0,start) + rep + textarea.value.substring(end,len);		
    			textarea.scrollTop = scrollTop;
    			textarea.scrollLeft = scrollLeft;
    		}
    }

    function doList(tag1,tag2,obj)
    {
    	textarea = document.getElementById(obj);	
    	if (document.selection) // Code for IE
    	{
    	  textarea.focus();
    		var sel = document.selection.createRange();
    		var list = sel.text.split('\n');
    		for(i=0;i<list.length;i++) 
    		{ list[i] = '* ' + list[i]; }
    		sel.text = list.join("\n");
    	} 
    	else // Code for Firefox
    	{
    		var len = textarea.value.length;
    	   var start = textarea.selectionStart;
    		var end = textarea.selectionEnd;
    		var i;
    		var scrollTop = textarea.scrollTop;
    		var scrollLeft = textarea.scrollLeft;
         var sel = textarea.value.substring(start, end);
    		var list = sel.split('\n');
    		for(i=0;i<list.length;i++) 
    		{ list[i] = '* ' + list[i]; }
    		var rep = list.join("\n");
    		textarea.value =  textarea.value.substring(0,start) + rep + textarea.value.substring(end,len);
    		textarea.scrollTop = scrollTop;
    		textarea.scrollLeft = scrollLeft;
     	 }
    }
    gEditor('geditor'); </script>
    )
  end
end

ActionView::Base.send :include, GeditorHelper